### ISSUE TITLE
Variables in bijectors cannot be reused.

### ISSUE THREAD

**Issue by shuiruge** [tensorflow/tensorflow#22052]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Centos 6
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.9.0 (tested in 1.10.0 too)
- **Python version**: 3.6
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: 8.0
- **GPU model and memory**: NVIDIA Tesla, 24GB
- **Exact command to reproduce**: the code within the "Source code / logs"

### Describe the problem
I am trying to reuse the weights and biases in the neural network within the MaskedAutoregressiveFlow bijector, by placing it within a `tf.variable_scope` with `reuse=tf.AUTO_REUSE`. But found that the weights and biases are not reused in practice.

### Source code / logs
```
import tensorflow as tf
from tensorflow.contrib.distributions.python.ops import bijectors as tfb

def get_bijector(name='my_bijector', reuse=None):
  """Returns a MAF bijector."""
  with tf.variable_scope(name, reuse=reuse):
    shift_and_log_scale_fn = \
        tfb.masked_autoregressive_default_template([128])
    return tfb.MaskedAutoregressiveFlow(shift_and_log_scale_fn)
  
x = tf.placeholder(shape=[None, 64], dtype='float32', name='x')

bijector_0 = get_bijector(reuse=tf.AUTO_REUSE)
y_0 = bijector_0.forward(x)

bijector_1 = get_bijector(reuse=tf.AUTO_REUSE)
y_1 = bijector_1.forward(x)

# We were expecting that the `y_0` and `y_1` share the same dependent variables,
# since we used `tf.AUTO_REUSE` within the `tf.variable_scope`. However, the following
# will return a `False`.
print(get_dependent_variables(y_0) == get_dependent_variables(y_1))
```

wherein we have employed the function that gains all the variables a tensor depends on:

```
import collections

def get_dependent_variables(tensor):
  """Returns all variables that the tensor `tensor` depends on.
  
  Forked from: https://stackoverflow.com/a/42861919/1218716
  
  Args:
    tensor: Tensor.
    
  Returns:
    List of variables.
  """  
  # Initialize
  starting_op = tensor.op
  dependent_vars = []
  queue = collections.deque()
  queue.append(starting_op)
  op_to_var = {var.op: var for var in tf.trainable_variables()}
  visited = {starting_op}

  while queue:
    op = queue.popleft()
    try:
      dependent_vars.append(op_to_var[op])
    except KeyError:
      # `op` is not a variable, so search its inputs (if any). 
      for op_input in op.inputs:
        if op_input.op not in visited:
          queue.append(op_input.op)
          visited.add(op_input.op)
          
  return dependent_vars
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by tensorflowbutler** [comment]
It has been 17 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by shuiruge** [comment]
> It has been 17 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

Yes.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
@jvdillon : Could you triage?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jvdillon: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jvdillon: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jvdillon: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sushreebarsa** [comment]
@shuiruge 
We see that you are using old version of tensorflow 1.x which is not actively supported, We recommend that you upgrade to 2.4 or later version.Please have a look at the [migration](https://www.tensorflow.org/guide/migrate) guide for reference to Migrate from TensorFlow 1.x to TensorFlow 2.Thank you!
